### PR TITLE
Tolerate bad encoding header in commit log message.  Fixed #179.

### LINKED ISSUE #179: help (java.nio.charset.IllegalCharsetNameException: utf-8,ISO-8859-1)
i use the bfg-repo-cleaner to clean my repo.
use this command : java -Dfile.encoding="UTF-8" -jar bfg.jar --private --strip-blobs-bigger-than 100M  NW_Router_V8_MAIN.git

But every time, I get this error:

`Found 169315 commits
Cleaning commits:        15% ( 25398/169315)Exception in thread "main" com.google.common.util.concurrent.UncheckedExecutionException: com.google.common.util.concurrent.UncheckedExecutionException: java.nio.charset.IllegalCharsetNameException: utf-8,ISO-8859-1
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2203)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3937)
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941)
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824)
        at com.madgag.git.bfg.MemoUtil$$anonfun$concurrentCleanerMemo$1$$anon$1.apply(memo.scala:61)
        at com.madgag.git.bfg.cleaner.ObjectIdCleaner.apply(ObjectIdCleaner.scala:75)
        at com.madgag.git.bfg.cleaner.RepoRewriter$$anonfun$clean$1$1$$anonfun$apply$mcV$sp$3.apply(RepoRewriter.scala:113)
        at com.madgag.git.bfg.cleaner.RepoRewriter$$anonfun$clean$1$1$$anonfun$apply$mcV$sp$3.apply(RepoRewriter.scala:112)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at com.madgag.git.bfg.cleaner.RepoRewriter$$anonfun$clean$1$1.apply$mcV$sp(RepoRewriter.scala:111)
        at com.madgag.git.bfg.cleaner.RepoRewriter$$anonfun$clean$1$1.apply(RepoRewriter.scala:104)
        at com.madgag.git.bfg.cleaner.RepoRewriter$$anonfun$clean$1$1.apply(RepoRewriter.scala:104)
        at com.madgag.git.bfg.Timing$.measureTask(timing.scala:40)
        at com.madgag.git.bfg.cleaner.RepoRewriter$.clean$1(RepoRewriter.scala:104)
        at com.madgag.git.bfg.cleaner.RepoRewriter$.rewrite(RepoRewriter.scala:145)
        at com.madgag.git.bfg.cli.Main$$anonfun$1.apply(Main.scala:59)
        at com.madgag.git.bfg.cli.Main$$anonfun$1.apply(Main.scala:34)
        at scala.Option.map(Option.scala:146)
        at com.madgag.git.bfg.cli.Main$.delayedEndpoint$com$madgag$git$bfg$cli$Main$1(Main.scala:33)
        at com.madgag.git.bfg.cli.Main$delayedInit$body.apply(Main.scala:27)
        at scala.Function0$class.apply$mcV$sp(Function0.scala:34)
        at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
        at scala.App$$anonfun$main$1.apply(App.scala:76)
        at scala.App$$anonfun$main$1.apply(App.scala:76)
        at scala.collection.immutable.List.foreach(List.scala:381)
        at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
        at scala.App$class.main(App.scala:76)
        at com.madgag.git.bfg.cli.Main$.main(Main.scala:27)
        at com.madgag.git.bfg.cli.Main.main(Main.scala)
Caused by: com.google.common.util.concurrent.UncheckedExecutionException: java.nio.charset.IllegalCharsetNameException: utf-8,ISO-8859-1
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2203)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3937)
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941)
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824)
        at com.madgag.git.bfg.MemoUtil$$anonfun$concurrentCleanerMemo$1$$anon$1.apply(memo.scala:61)
        at com.madgag.git.bfg.cleaner.ObjectIdCleaner$$anonfun$uncachedClean$1.apply(ObjectIdCleaner.scala:87)
        at com.madgag.git.bfg.cleaner.ObjectIdCleaner$$anonfun$uncachedClean$1.apply(ObjectIdCleaner.scala:85)
        at com.madgag.git.bfg.MemoUtil$$anon$3.load(memo.scala:75)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197)
        ... 28 more
Caused by: java.nio.charset.IllegalCharsetNameException: utf-8,ISO-8859-1
        at java.nio.charset.Charset.checkName(Charset.java:320)
        at java.nio.charset.Charset.lookup2(Charset.java:496)
        at java.nio.charset.Charset.lookup(Charset.java:475)
        at java.nio.charset.Charset.forName(Charset.java:540)
        at org.eclipse.jgit.util.RawParseUtils.parseEncoding(RawParseUtils.java:791)
        at org.eclipse.jgit.revwalk.RevCommit.getEncoding(RevCommit.java:529)
        at com.madgag.git.bfg.model.CommitNode$.apply(Commit.scala:62)
        at com.madgag.git.bfg.model.Commit$.apply(Commit.scala:33)
        at com.madgag.git.bfg.cleaner.ObjectIdCleaner$$anonfun$1.apply(ObjectIdCleaner.scala:100)
        at com.madgag.git.bfg.cleaner.ObjectIdCleaner$$anonfun$1.apply(ObjectIdCleaner.scala:98)
        at com.madgag.git.bfg.MemoUtil$$anon$3.load(memo.scala:75)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197)
        ... 39 more`
I've actually ran bfg on another repository which was much bigger too. Also, I've just updated Java with no luck.
@rtyley

### PR DESCRIPTION
**Pull request by javabrett**

### CODE REVIEW

**Comment by javabrett** [comment]
Fixed #179.

File: bfg-library/src/main/scala/com/madgag/git/bfg/model/Commit.scala [hunk 1]
```diff
@@ -59,7 +60,8 @@ case class CommitArcs(parents: Seq[ObjectId], tree: ObjectId) {
 }
 
 object CommitNode {
-  def apply(c: RevCommit): CommitNode = CommitNode(c.getAuthorIdent, c.getCommitterIdent, c.getFullMessage, c.getEncoding)
+  def apply(c: RevCommit): CommitNode = CommitNode(c.getAuthorIdent, c.getCommitterIdent, c.getFullMessage,
+      try c.getEncoding catch {case e: IllegalCharsetNameException => Constants.CHARSET})
```
Comment (rtyley) [inline bfg-library/src/main/scala/com/madgag/git/bfg/model/Commit.scala]: Possibly a refinement would be to split on commas and take the first encoding that we can decode - which might be different to the `UTF-8` of `Constants.CHARSET` - but so far as I can tell, nobody has ever expected the `encoding` header in Git commit messages to contain commas. The implementation in JGit certainly doesn't expect it:

https://github.com/eclipse/jgit/blob/ccc89977/org.eclipse.jgit/src/org/eclipse/jgit/util/RawParseUtils.java#L764-L800

...and C-Git ultimately uses `iconv`, which doesn't seem to expect comma separated values either: http://www.gnu.org/savannah-checkouts/gnu/libiconv/documentation/libiconv-1.13/iconv_open.3.html

File: bfg/src/test/scala/com/madgag/git/bfg/cli/MainSpec.scala [hunk 2]
```diff
@@ -110,6 +110,12 @@ class MainSpec extends Specification {
         run("--delete-folders bar --delete-files foo.txt")
       }
     }
+    
+    "not crash on encountering a commit with bad encoding header" in new unpackedRepo("/sample-repos/badEncoding.git.zip") {
```
Comment (rtyley) [inline bfg/src/test/scala/com/madgag/git/bfg/cli/MainSpec.scala]: The important commit object within this repo:

```
$ git cat-file -p 0554b30af831719d1de53ee2ee8ff18d3f637953
tree 2b297e643c551e76cfa1f93810c50811382f9117
author Brett Randall <javabrett@gmail.com> 1476143356 +1100
committer Brett Randall <javabrett@gmail.com> 1476143356 +1100
encoding utf-8,ISO-8859-1

This commit message will have a bad Encoding header of Encoding utf-8,ISO-8859-1
```

... the important line being `encoding utf-8,ISO-8859-1` - ie an invalid encoding.

**Comment by rtyley** [comment]
Great work @javabrett - thank you!

**Pull request merged**